api/sms_provider/sms_provider_test.go

```diff
@@ -0,0 +1,83 @@
+package sms_provider
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+	"testing"
+
+	"github.com/netlify/gotrue/conf"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+var handleApiRequest func(*http.Request) (*http.Response, error)
+
+type SmsProviderTestSuite struct {
+	suite.Suite
+	Config *conf.GlobalConfiguration
+	Client HttpClient
+}
+
+type MockHttpClient struct {
+	mock.Mock
+}
+
+func (m *MockHttpClient) Do(req *http.Request) (*http.Response, error) {
+	return handleApiRequest(req)
+}
+
+func TestSmsProvider(t *testing.T) {
+
+	ts := &SmsProviderTestSuite{
+		Config: &conf.GlobalConfiguration{
+			Sms: conf.SmsProviderConfiguration{
+				Twilio: conf.TwilioProviderConfiguration{
+					AccountSid:        "test_account_sid",
+					AuthToken:         "test_auth_token",
+					MessageServiceSid: "test_message_service_id",
+				},
+			},
+		},
+		Client: &MockHttpClient{},
+	}
+	suite.Run(t, ts)
+}
+
+func (ts *SmsProviderTestSuite) TestTwilioSendSms() {
+	twilioProvider, err := NewTwilioProvider(ts.Config.Sms.Twilio)
+	require.NoError(ts.T(), err)
+
+	client = ts.Client
+
+	handleApiRequest = func(req *http.Request) (*http.Response, error) {
+		// check authorization header
+		require.Contains(ts.T(), req.Header, "Authorization")
+
+		// check request body sent
+		require.NoError(ts.T(), req.ParseForm())
+		require.Contains(ts.T(), req.Form, "To")
+		require.Contains(ts.T(), req.Form, "Channel")
+		require.Contains(ts.T(), req.Form, "From")
+		require.Contains(ts.T(), req.Form, "Body")
+
+		b, err := json.Marshal(&SmsStatus{
+			To:     req.Form["To"][0],
+			From:   req.Form["From"][0],
+			Status: "sent",
+			Body:   req.Form["Body"][0],
+		})
+		require.NoError(ts.T(), err)
```

Comment:
Shouldn't we check the params passed to a provider? Both from the provider config and request?

Reply:
hmm what do you mean by the params passed to a provider? the checks from line 60-64 already check if the request made to a twilio api endpoint has the right params - is there something i'm missing here?

Reply:
I thought it only checks the existence of the params but not the values. Am I missing something?

Reply:
ah do you mean we should also check the values passed [here](https://github.com/supabase/gotrue/blob/master/api/sms_provider/twilio.go#L59-L62) to make sure that if they are modified, the test fails?

Reply:
Yep! Instead of checking that form field “To” exists, check the the passed phone number.

Reply:
decided to use [gock](https://github.com/h2non/gock) instead! it allows me to mock the external requests without having to create an interface and they also support request input validation